include/translatable-object.php

```diff
@@ -0,0 +1,463 @@
+			} elseif ( $store_only_term_ids ) {
+				$to_cache = array( $terms[ $tax ]->term_id );
+			} else {
+				// Backward compatibility with WP < 6.0.
+				$to_cache = array( $terms[ $tax ] );
+			}
+
+			wp_cache_add( $id, $to_cache, "{$tax}_relationships" );
+		}
+
+		return $term;
+	}
+
+	/**
+	 * A JOIN clause to add to sql queries when filtering by language is needed directly in query.
+	 *
+	 * @since 3.4
+	 *
+	 * @param string $alias Optional alias for object table.
+	 * @return string The JOIN clause.
+	 *
+	 * @phpstan-return non-empty-string
+	 */
+	public function join_clause( $alias = '' ) {
+		global $wpdb;
+
+		if ( empty( $alias ) ) {
+			$alias = $this->db['default_alias'];
+		}
+
+		return " INNER JOIN {$wpdb->term_relationships} AS pll_tr ON pll_tr.object_id = {$alias}.{$this->db['id_column']}";
+	}
+
+	/**
+	 * A WHERE clause to add to sql queries when filtering by language is needed directly in query.
+	 *
+	 * @since 1.2
+	 *
+	 * @param PLL_Language|PLL_Language[]|string|string[] $lang A `PLL_Language` object, or a comma separated list of language slugs, or an array of language slugs or objects.
+	 * @return string The WHERE clause.
+	 *
+	 * @phpstan-param PLL_Language|PLL_Language[]|non-empty-string|non-empty-string[] $lang
+	 */
+	public function where_clause( $lang ) {
+		/*
+		 * $lang is an object.
+		 * This is generally the case if the query is coming from Polylang.
+		 */
+		if ( $lang instanceof PLL_Language ) {
+			return ' AND pll_tr.term_taxonomy_id = ' . absint( $lang->get_tax_prop( $this->tax_language, 'term_taxonomy_id' ) );
+		}
+
+		/*
+		 * $lang is an array of objects, an array of slugs, or a comma separated list of slugs.
+		 * The comma separated list of slugs can happen if the query is coming from outside with a 'lang' parameter.
+		 */
+		$languages        = is_array( $lang ) ? $lang : explode( ',', $lang );
+		$languages_tt_ids = array();
+
+		foreach ( $languages as $language ) {
+			$language = $this->model->get_language( $language );
+
+			if ( ! empty( $language ) ) {
+				$languages_tt_ids[] = absint( $language->get_tax_prop( $this->tax_language, 'term_taxonomy_id' ) );
+			}
+		}
+
+		if ( empty( $languages_tt_ids ) ) {
+			return '';
+		}
+
+		return ' AND pll_tr.term_taxonomy_id IN ( ' . implode( ',', $languages_tt_ids ) . ' )';
+	}
+
+	/**
+	 * Returns the IDs of the objects without language.
+	 *
+	 * @since 3.4
+	 * @throws Exception When implementing PLL_Translatable_Object_With_Types_Interface without providing the field
+	 *                   'type_column' in the class property `$db`.
+	 *
+	 * @param int $limit Max number of objects to return. `-1` to return all of them.
+	 * @return int[] Array of object IDs.
+	 *
+	 * @phpstan-param -1|positive-int $limit
+	 * @phpstan-return list<positive-int>
+	 */
+	public function get_objects_with_no_lang( $limit ) {
+		global $wpdb;
+
+		$object_types_sql = '';
+
+		if ( $this instanceof PLL_Translatable_Object_With_Types_Interface ) {
```

Comment:
`PLL_Translatable_Object` cannot implement `PLL_Translatable_Object_With_Types_Interface`, it would force the child classes to implement the methods `get_translated_object_types()` and `is_translated_object_type()`, which is what we try to prevent with this interface.

Reply:
In my mind `get_translated_object_types()` could return an empty array and `is_translates_object_type()` always `false` in such case.
But I understand your point.

Reply:
I don't understand the idea of moving this method into a trait, since it is also required when there is no sub-types.

Reply:
Fixed in [0549fb5](https://github.com/polylang/polylang/pull/1159/commits/0549fb514e875b9c555515fb32c62713fe64d5cb).